File: examples/GaussianSmear.cpp
#include <Gaussian.h>
#include <Convolution.h>
#include <BinnedPdf.h>
#include <PdfConverter.h>

int main(){
    
    // Test it on a binned pdf
    Gaussian testGaus(std::vector<double>(2,1), std::vector<double>(2,2)); // means then sigmas
    AxisCollection axes;
    axes.AddAxis(PdfAxis("energy", 2, 3, 10, "Energy"));
    axes.AddAxis(PdfAxis("radius", 2, 3, 10, "rad"));

    BinnedPdf testPdf = PdfConverter::ToBinnedPdf(testGaus, axes);

    Gaussian gausPdf(1, 2); // mean sigma
    Convolution conv;

    conv.SetPdf(&gausPdf);
    conv.SetAxes(axes);

    // What indicies matter?
    std::vector<size_t> smearDimensions(1, 0); // just first index for smear
    DataRepresentation  smearRep(smearDimensions);

    // pdfs to act on will be in E and r
    smearDimensions.push_back(1);
    DataRepresentation pdfRep(smearDimensions);

    conv.SetDataRep(smearRep);
    conv.SetPdfDataRep(pdfRep);
    
    // Build the internal matrix
    conv.Construct();

    // Now use it 
    BinnedPdf transformed = conv(testPdf);
                             
    return 0;
}

File: examples/gaussian_smear.py
import oxsx
import matplotlib.pyplot as plt

if __name__=="__main__":

    # Have to pass a custom list type for now 
    # (see http://stackoverflow.com/questions/15842126 )
    v1 = [1.0, 1.0]
    v2 = [2.0, 2.0]
    test_gaus = oxsx.Gaussian([1.0, 1.0], [2.0, 2.0])

    axis1 = oxsx.PdfAxis("energy", 2, 3, 10, "Energy")
    axis2 = oxsx.PdfAxis("radius", 2, 3, 10, "rad")
    axes = oxsx.AxisCollection()
    axes.add_axis(axis1)
    axes.add_axis(axis2)

    test_pdf = oxsx.PdfConverter.to_binned_pdf(test_gaus, axes)

    gaussian = oxsx.Gaussian(1, 2) # Must always pass lists
    conv = oxsx.Convolution()
    conv.set_axes(axes)
    conv.set_function(gaussian)

    smear_rep = oxsx.DataRepresentation([0])
    pdf_rep = oxsx.DataRepresentation([0, 1])

    conv.set_data_rep(smear_rep)
    conv.set_pdf_data_rep(pdf_rep)

    print gaussian.integral()

    conv.construct()

    transformed = conv(test_pdf)

    # Still need to transform from DoubleList to an actual list...
    # Not sure how to extract the axis information.
    contents = test_pdf.get_bin_contents()
    plt.plot(range(len(contents)), contents, 'bo') 
    plt.show()
